### DISCUSSION TITLE
Orange Pi 5 Plus:  Is it possible to change fan behavior? (fan 2pin header)

### DISCUSSION THREAD

**Discussion by TuTuc0** [Joshua-Riek/ubuntu-rockchip#311] [Q&A]
Currently fan plugged into fan header behaves according cpu temperature defined in hex table located in /sys/firmware/devicetree/base/pwm-fan/rockchip,temp-trips  
and powering the fan with power from this hex table  /sys/firmware/devicetree/base/pwm-fan/cooling-levels

So it behaves like:   above 50c - power 50, above 55c - power 100, ..., ,above 70c - power 255

Is it possible to somehow set the fan to be always at same power level? Or even adjust the CPU cooling tables?

It is possible to temporary set fan speed by   `echo 80 > /sys/class/hwmon/hwmon9/pwm1`  however speed is overwritten when another cpu temperature level is reached.

**Comment by levinhtin** [comment]
The same question

**Comment by stevepringle** [comment]
See my post for the rock 5a board. That should help.
https://github.com/Joshua-Riek/ubuntu-rockchip/discussions/469

**Reply by TuTuc0** [reply]
Thanks, finally got it working on my OPI5+. With your help and ChatGPT I tweaked it a little and it worked. 
In your mentioned post, there is a typo in command I think;
`sudo nano /boot/firmware/ubuntuEnv.txtuntuEnv.txt` should be `sudo nano /boot/firmware/ubuntuEnv.txt` instead?


For my OPI5+ and 5V Noctua fan, I have used:
```
/dts-v1/;
/plugin/;
/ {
        fragment@0 {
                target = <&fan>;
                __overlay__ {
                        cooling-levels = <80 80 80 90 90 120>;
                };
        };
};

```

**Reply by stevepringle** [reply]
Thanks, I've updated the post to remove the typo :-)

**Comment by StuartIanNaylor** [comment]
Its sort of a misnomer that you can control 2 pin fans by pwm, that quickly gets quite complex.
A 2 pin fan is a BLDC fan motor that has an embedded controller multipole which means 3 or more.
3 pin fans are a 2 pin fan with a tacho output and voltage controlled but the manufacturer expects them to be voltage controlled so the control circuit often works over a extended voltage range.
4 pin 4th pin is pwm duty cycle using the same 3rd pin tacho.

So yeah you can use PWM on a 2pin fan but the fan curve is likely pretty bad starting at 40/45% and likely if you could measure the speed having little correlation to pwm duty as the PWM is just aproximating voltage.
I have no idea where the idea came from for using 2pin fans on a transistor on a gpio pwm other than because they could and it was cheaper than fairly cheap 4 pin fans or 3 pin gpio fans.

The later 3 pin gpio fan is sort of the newer one to market as its a 4 pin fan with the 3rd wire cut off :)
They will stall lower than 25% but the duty cycle will correlate to speed and use a period above hearing 20-25Khz.
That is how pwm fans work and you will get a proper linear output all the way to 100% and not just a series of jumps.
https://github.com/StuartIanNaylor/Opi5PWM-fan

20mm 3 pin gpio fan https://www.raspberrypi.com/products/raspberry-pi-4-case-fan/
30mm 3 pin gpio fan https://thepihut.com/products/software-controllable-5v-30mm-fan-for-raspberry-pi
40mm 3 pin gpio fan https://www.aliexpress.com/item/1005002518918026.html

They do have a much better and more controllable fan curve as they are 5v fans that will take a 3.3v pwm for the duty cycle.

Yeah you can do the above and it will work in a fashion or just get a gpio fan and ignore that stupid 2pin header as surely there was space for a 2.54mm 3pin connector somewhere and if not used you should be able to steal back a gpio and gain another 5vcc & Gnd pins...
